plugins/guests/linux/cap/change_host_name.rb

```diff
@@ -0,0 +1,38 @@
+require 'vagrant/util/guest_hosts'
+
+module VagrantPlugins
+  module GuestLinux
+    module Cap
+      module ChangeHostName
+        include Vagrant::Util::GuestHosts::Linux
+
+        def change_name_command(name)
+          return <<-EOH.gsub(/^ {14}/, '')
+              # Set the hostname
+              echo '#{name}' > /etc/hostname
+              hostname '#{name}'
+            EOH
+        end
+
+        def change_host_name?(comm, name)
+          !comm.test("hostname -f | grep '^#{name}$'", sudo: false)
+        end
+
+        def change_host_name(machine, name)
+          comm = machine.communicate
+          
+          network_with_hostname = machine.config.vm.networks.map {|_, c| c if c[:hostname] }.compact[0]
+          if network_with_hostname
+            replace_host(comm, name, network_with_hostname[:ip])
+          else
+            add_hostname_to_loopback_interface(comm, name)
+          end
+
+          if change_host_name?(comm, name)
+            comm.sudo(change_name_command(name))
+          end
+        end
```

Comment:
Wrap these methods in a module, add an `.extended` method, and then extend in the `ChangeHostName` module:

```ruby
module ChangeHostName
  module Methods
    def change_name_command(name)
      ...
    end
    ...
    def self.extended(klass)
      klass.extend(Vagrant::Util::GuestHosts::Linux)
    end
  end

  extend Methods
end
```